### DISCUSSION TITLE
Add a "shading" language to Audio nodes, to create sounds effects on top of streams

### DISCUSSION THREAD

**Discussion by vPumpking** [godotengine/godot-proposals#6476] [Engine Core]
### Describe the project you are working on

it's called Godot Engine 😉
just searching how to increase the Godot's power

### Describe the problem or limitation you are having in your project

the list of possible effects for sounds is very short

### Describe the feature / enhancement and how it helps to overcome the problem or limitation

Add a thing like shading language to enable to freely post-effect the streams of audio and video players

### Describe how your proposal will work, with code, pseudo-code, mock-ups, and/or diagrams

A bit like in a `canvas item` or in a `spatial` node, the user can open an `Effects` section in the inspector when an Audio or video player is selected. In this section, the user can create a new "`AudioShader`" file, using a new language like GLSL to post-effect the stream output.

### If this enhancement will not be used often, can it be worked around with a few lines of script?

no

### Is there a reason why this should be core and not an add-on in the asset library?

it's too hard with an addon

**Comment by Calinou** [comment]
Moving to discussion, as there is no concrete technical implementation provided.

**Reply by vPumpking** [reply]
ok

**Comment by theraot** [comment]
ShaderToy does something similar. This is the default example:

```
vec2 mainSound( int samp, float time )
{
    // A 440 Hz wave that attenuates quickly overt time
    return vec2( sin(6.2831*440.0*time)*exp(-3.0*time) );
}
```

*I could not figure out what `samp` is.*

The result is a `vec2` where the `x` and `y` components are `left` and `right` audio channel for stereo sound.

Yes, it uses the same language that they use for graphics.

I don't know how it works behind the scenes. However, I can imagine that using the GPU to compute an audio buffer this way could use a fragment shader with time passed as a varying uniform and render to a texture, or whatever the equivalent is using compute shaders. From there the buffer would go the audio pipeline.

**Reply by theraot** [reply]
I have managed to make a proof of concept. It is working with a fragment shader in Godot 3.5. However, I have audio popping. It is something with timing, the viewport render delay, and stitching the buffer. I know it is not working correctly, but proof of concept is proof of concept.

Edit: I forgot to mention that it is playing twice, and it shouldn't.

[GPU2Audio.zip](https://github.com/godotengine/godot-proposals/files/10948852/GPU2Audio.zip)

**Reply by vPumpking** [reply]
It gets good results! I suggest we continue to explore this way

**Reply by theraot** [reply]
Fixed: [GPU2Audio.zip](https://github.com/godotengine/godot-proposals/files/10949503/GPU2Audio.zip)

It is also a lesson in using the `AudioStreamGeneratorPlayback` class. Instead of calling `_fill_buffer` from `_process` (which is what the official demo does), I'm using `call_deferred("_fill_buffer")`, which is what ultimately fixed the audio popping. The other changes making sure I don't try to generate more samples than I fit in a frame, and that I make the next frame from the exact next sample after the last one.

**Reply by theraot** [reply]
Optimization: [GPU2Audio.zip](https://github.com/godotengine/godot-proposals/files/10949598/GPU2Audio.zip)

This version is using the four color channels to generate two audio frames per pixel.

I also did some renaming.

I tried different approaches to try to pass the data from the texture to the audio buffer. But the naive approach seems to be the faster way to do it in GDScript.

**Reply by sairam4123** [reply]
int samp could mean sampling of sound, though it's confusing, it's most likely audio sampling

**Reply by somebody1234** [reply]
`samp` seems to be an `int`eger sample number (a count starting from 0), as an alternative to the `float` time value

**Comment by Akosmo** [comment]
If I understand this right? You want an audio "shading" language that will receive audio sample output from buses, modify them, and return to those same buses (or whatever buses they are routed to)?

Why not use GDExtesion to make use of `AudioEffect`, `AudioEffectInstance`, and `AudioFrame`? I think creating a language would add too much to the engine, it'd be very complex, hard to maintain, and not many people would use it. You can already do that entire process, it's basically how effects are done already.

Do you have any examples that justifies this proposal? I'm curious
